### ISSUE TITLE
DIY Pen Plotter with Automatic Tool Changer | CNC Drawing Machine

### ISSUE THREAD

**Issue by yifandianzi** [svenhb/GRBL-Plotter#361]
![IS7P0{A9 K{LO IH4V_C0IF](https://github.com/svenhb/GRBL-Plotter/assets/94118847/68723f19-c294-4293-8173-1b98fc7bdfa5)

May I ask how to set the tool table above the Y-axis? After setting, the pen position in the tool table is negative, and all the tool tables in the pictures are in the negative direction of the Y-axis

**Comment by svenhb** [comment]
Tool table positions must be set in machine coordinates - homig of the plotter is needed (or manual "homig" via hard reset on lower left position):
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/79de4e34-4f4e-478d-8d21-36b91feaa6e1)

![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/63086945-a189-4786-895a-87ce0228d8ec)

**Comment by svenhb** [comment]
Just change the 277 to 0

**Comment by yifandianzi** [comment]
![H%09TL5$IN20}Y}6D9 GJPB](https://github.com/svenhb/GRBL-Plotter/assets/94118847/0d4decd1-a6c4-4717-b9f2-b9728e955bae)

https://github.com/svenhb/GRBL-Plotter/assets/94118847/46c40b4b-b3ee-48c3-ac87-54c015b7c436

Hello, this is a demonstration set to 0

**Comment by svenhb** [comment]
You are right, there is a bug and the offset is not shown, but processed during streaming. Will fix it soon.

**Comment by svenhb** [comment]
You may try this version: https://github.com/svenhb/GRBL-Plotter/blob/master/GRBL-Plotter_Setup_1_7_0_2_i.exe

**Comment by yifandianzi** [comment]
https://github.com/svenhb/GRBL-Plotter/assets/94118847/1228b2b6-13a1-491b-bb9c-f4277c031af0

Hello, there is another issue now, which is that it can return to its normal position when returning to zero. When starting to work, the Z motor will rotate in the opposite direction, and then it can surprise the next step normally

**Comment by svenhb** [comment]
I don't exactly understand this issue. I assume the problem is the servo-position, when PWM is off.
Compare behavior of PWM on/off here:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/63e467a3-fb55-415f-affa-fc2f7a511daf)

Compare behavior of PWM low/high here:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/5079138b-c86c-4de4-be4d-3a38e745bd22)

**Comment by svenhb** [comment]
Now I could watch your video on a PC - it's not a servo...
I thing you need to "Zero Z" your Z axis before starting streaming.

**Comment by yifandianzi** [comment]
![BD$3O7WH~ 093QO9ZLO_5QG](https://github.com/svenhb/GRBL-Plotter/assets/94118847/c3700e14-3ae7-4257-a78e-51a80af0d203)
Hello, I have reset all X, Y, and Z here to zero, but when starting work, they still rotate in the opposite direction. I think it may be due to incorrect settings elsewhere, and it should not be rotated without selecting a pen

**Comment by svenhb** [comment]
It is normal that the Z axis move to a save position on start.
Example code after importing a simple line:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/3abd76e0-2cdf-46ea-95a1-fce311ffb53c)

Line 14 G00 Z2.00 move Z upwards to save position
Line 17 move to start position
Line 18 G01 Z-1 move Z down, to start drawing
Line 19 draw the line
Line 20 G00 Z2.00 mive Z upwards

**Comment by yifandianzi** [comment]
![7S7WXMXO~T9A%)U84`9ORX0](https://github.com/svenhb/GRBL-Plotter/assets/94118847/75da3bfb-43ad-4a25-b8e4-58e64959c6e5)
![XG43{Y)IQRDA_2H)(D%O84L](https://github.com/svenhb/GRBL-Plotter/assets/94118847/899dec1e-23ec-4873-9af9-0899e73979b2)
Did I make the right changes here

**Comment by svenhb** [comment]
I don't know your hardware, I didn't saw a servo for a gripper...
Do your gripper opens with M3 S100?

Did you check my site? https://grbl-plotter.de/index.php?id=auto-pen-change&setlang=en

**Comment by svenhb** [comment]
Looks like the Z-stepper should turn for a half turn maximum to move from lowest Z to highest Z.
Did you setup $102 correctly?

**Comment by yifandianzi** [comment]
Should I set $102 correctly to $40?
https://www.youtube.com/watch?v=ZxeStwRespc&list=RDCMUCJHZY4Tl6ovsnBrNocdDbAg&start_radio=1&rv=ZxeStwRespc&t=65
https://www.youtube.com/watch?v=The following link is mine_4S9PE

**Comment by svenhb** [comment]
Last video is not visible. I don't know your driver settings for micro-steps.
Usually a stepper-motor needs 200 steps for a full turn, If I interpret your design correctly, you have a half turn for about 20mm Z-change. With full-step-mode you may set $102 = 5 (100 steps/half-turn divided by 20 mm)
If you have half-step mode: $102 = 10

If your Z axis is spring loaded to keep it up, you may have a problem with step loss at the Z motor.

**Comment by yifandianzi** [comment]
https://www.youtube.com/watch?v=ZJZuB_4S9PE
Hello, setting $102=10 and $102=5 will prompt an error, while setting $102=40 will not cause an error

**Comment by yifandianzi** [comment]
I have set these three places in the firmware
![RA4K2ELA)IJ@}4A)US2) E7](https://github.com/svenhb/GRBL-Plotter/assets/94118847/a04cdb6e-e076-4d23-9ed9-9f4f87459f55)

Do you think it's correct

**Comment by svenhb** [comment]
I don't understand how your z-axis works, in your video time 0.04, the stepper is moving, but the z-wagon is not moving....
Also I don't see an end-switch for z. So you can't use homing for z.
You need to set: #define HOMING_CYCLE 0 ((1<<X_AXIS) | (1<<Y_AXIS)) if you have end-switches there

**Comment by yifandianzi** [comment]
https://www.youtube.com/watch?v=NDZsdDLbDDo
Normally, it should start working after resetting to zero, and then X Y starts moving to the pen selection position and Z starts lifting. The problem now is that when the start button is clicked, the Z stepper motor has already started working, and then the Z limit is set off

**Comment by yifandianzi** [comment]
https://www.youtube.com/watch?v=ZxeStwRespc&list=RDCMUCJHZY4Tl6ovsnBrNocdDbAg&start_radio=1&rv=ZxeStwRespc&t=65
The author's point is that the Z-lift motor does not work when starting work after resetting to zero

**Comment by svenhb** [comment]
The main problem is, that pen selection positions are machine coordinate, which needs homing.
Also the linked video shows a working z-homing - you need to change your mechanics...
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/e968dfd5-4196-47fe-bce6-2371bc7d4ffc)

**Comment by yifandianzi** [comment]
Thank you for your great help. When you pointed out the location, I also raised the Z limit position to zero. Now, just as I was preparing to start the work, the Z stepper motor continued to raise and then raised again, triggering a limit fault

**Comment by yifandianzi** [comment]
https://www.youtube.com/watch?v=QQNdMw_KsNo
After all X Y Z are reset, the execution starts at 0:53, and the Z stepper motor continues to rise

**Comment by svenhb** [comment]
After Z-homing, Z stays at top position which you may want to have at +2mm in work coordinates.
After Z homing move 2mm down and press "Zero Z".
Also important the choosen travel way for import - small enough for your hardware?:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/1e111fb7-d7b3-4ee6-a498-1f0124125291)

**Comment by yifandianzi** [comment]
![VO{S~}7XIUGC@X%3%V(9Q%O](https://github.com/svenhb/GRBL-Plotter/assets/94118847/c7392e03-fadf-4f28-8847-e83038ba07f8)
Hello, where is the white area set and where can I reach the coordinates

**Comment by yifandianzi** [comment]
https://www.youtube.com/watch?v=qdFgPm4bOkc
This is my complete operation steps, but there are still problems and I don't have a clue now

**Comment by svenhb** [comment]
Looks good so far.
1. set your travel ranges via $130, 131, 132. https://github.com/gnea/grbl/wiki/Grbl-v1.1-Configuration#130-131-132--xyz-max-travel-mm
This will influence your homing machine coordinates:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/8e65a311-a84d-4513-8bf1-9ef013dff6e8)

After this you go to setup and take over the ranges:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/b3e86961-7436-41dd-b739-d770155e6bdf)

**Comment by svenhb** [comment]
Here you need a value, close to your homing Y-value = ca. 350:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/cdb66b4e-2819-4b1b-a32e-e9addf16847f)

**Comment by yifandianzi** [comment]
![2K9V_NV9%6PZGA SARUG{4K](https://github.com/svenhb/GRBL-Plotter/assets/94118847/380090be-badf-40c0-8df4-b2e8e5da547f)

Hello, where else do I need to set the Z stepper motor to continue rising after all XYZ resets? I set it to zero in the picture

**Comment by svenhb** [comment]
The problem is here:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/532c96c9-9aad-4e87-9403-8edd2f692d9f)
Y-Offset is 191, Tool-2 Y-pos is 172, so the machine tries to move to Y=(191+172)=363 - out of your limit.

You can check this box, to see sent commands:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/6f4c7641-0dfa-45b0-9bde-706a29d1585d)

**Comment by svenhb** [comment]
Perhaps you are not using the latest bug fix, with the tool-offset:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/27bd2cbc-6db1-47a8-8388-0f6f14f37b57)

https://github.com/svenhb/GRBL-Plotter/blob/master/GRBL-Plotter_Setup_1_7_0_2_i.exe

**Comment by yifandianzi** [comment]
Hello, this is my demonstration process. The X-axis has stopped before reaching the position of the pen holder. Do you need to set the X position in the picture
https://www.youtube.com/watch?v=qE-if9f868A
![09E0~L9AR(4%FB@KFB@ABKW](https://github.com/svenhb/GRBL-Plotter/assets/94118847/41f2087b-d895-48a1-a28b-796de70f518d)

**Comment by svenhb** [comment]
>The X-axis has stopped before reaching the position of the pen holder

Your homing x machine coordinate is 348:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/351cf704-d8d3-4930-a05a-5d586dc33caf)

For pen change, the the X-pos will be used:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/e155ce03-7700-4243-b892-002a2c086d0d)

So the plotter is moving to machine-X =0 at minute 0.45:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/6bd27aa5-4b6c-4be2-a792-31d54b3a56db)

Note: from here it moves 18mm into the pen-holder because of the script:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/932b5d35-6845-48f6-a36e-70619f7a386c)

Either you add an offset here: 
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/2bacb3f5-f1c1-4968-8642-22dbd10b4509)

Or you adapt the grbl travel range $130 https://github.com/gnea/grbl/wiki/Grbl-v1.1-Configuration#130-131-132--xyz-max-travel-mm

To avoid drawing on pen-holder:
Move your plotter manually (with joystick) to this point and press "Zero XY"
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/fb738796-57ac-426e-9f5d-4d23df6bd3cb)

**Comment by yifandianzi** [comment]
Thank you for your help. Now we can move on. Another question is: the graph to be drawn and the text to be written start from XY zero point. How can we move to the center of the plotter to draw? There are also tool tables 1, 2, 3, 4, 5.... Where should these positions be adjusted?
![B5YL2KCK6SJ(%B2LJQX8BLN](https://github.com/svenhb/GRBL-Plotter/assets/94118847/ea51e591-4857-44b2-8d7b-b2a263e6efc5)

**Comment by svenhb** [comment]
>How can we move to the center of the plotter to draw? 

Set up your 0;0 on the green cross (as shown above)
Load a graphic with Offset removal
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/2da8e5c9-d1ac-4814-b35b-a6d1f008fbff)

Then set your center coordinates here (e.g. 180;90) and "Apply Offset":
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/b8caff89-71fe-4855-a196-d2901cc84e36)



>There are also tool tables 1, 2, 3, 4, 5.... Where should these positions be adjusted?

Just measure your pen holders...
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/42daa33a-3c43-4a70-bacf-422868b20d75)

**Comment by yifandianzi** [comment]
Hello, the problem I am currently facing is with the software, which cannot be displayed at all. I tried to install the software by changing the computer system, but after opening it, it still remains the same. It cannot be maximized, and some settings also cannot see the screen    Using the latest version
https://www.youtube.com/watch?v=15d6JGEen88
![_9)N~ 2`}`9(FK1(7$JZTL](https://github.com/svenhb/GRBL-Plotter/assets/94118847/75f2a921-b154-417d-b197-361161ad0d6c)

**Comment by svenhb** [comment]
Should look like this (https://github.com/svenhb/GRBL-Plotter/blob/master/GRBL-Plotter_Setup_1_7_0_2_i.exe):
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/8c1526e6-9189-43e8-b8de-94cf62178d01)

![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/07db2068-ac2b-45ac-aba2-359c32499d1a)

**Comment by svenhb** [comment]
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/84cbc360-b86b-437c-8ce8-df9ab8f6f013)

**Comment by yifandianzi** [comment]
Hello, do you still need to set up the process of picking up and lifting the pen
https://www.youtube.com/watch?v=uYTOFpzlZCA

**Comment by svenhb** [comment]
Looks good so far.

- You may adjust the servo positions for open and close - see https://www.youtube.com/watch?v=uFfjXCRGSnM
- You may disable PWM-off when finished, to avoid strange servo position:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/eceb4ad2-d7e1-462f-94bf-4ffcef73804f)

**Comment by yifandianzi** [comment]
https://www.youtube.com/watch?v=WWQc27jp7ks
Thank you. After picking up the pen, the servo released it and it got stuck while writing. Where can I cancel this step

**Comment by svenhb** [comment]
I don't know where the last movement comes from.
- from footer setup?
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/962837de-374f-4872-9951-631566c3ebe3)

Switch on this check box and watch sent commands:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/f7655a35-d18a-4d8b-8a83-93b1ae58b5db)

**Comment by yifandianzi** [comment]
Hello, I have tried to modify these places, but they still don't work and I will loosen the pen
![${2JXW3A}6C6OCQ}AY52TH0](https://github.com/svenhb/GRBL-Plotter/assets/94118847/1514b76f-735f-40f9-b441-6085fee2fe52)

**Comment by yifandianzi** [comment]
In this way
![95)XUC$NS 5DJN}}O7{D07E](https://github.com/svenhb/GRBL-Plotter/assets/94118847/dcda6f25-da91-4c70-9c4f-0e7698924be0)

**Comment by yifandianzi** [comment]
https://www.youtube.com/shorts/gHtDPaVl4kA
It's Like That

**Comment by yifandianzi** [comment]
I am using a special version of firmware with servo support
https://github.com/cprezzi/grbl-servo

**Comment by svenhb** [comment]
I made a mistake in my last post...You need to use Z axis and disable spindle use.
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/e8095968-71a3-45ef-8d89-9f7082b4a177)

**Comment by yifandianzi** [comment]
Hello, is that the case
![T(VJ`L3JGP~RX$4JH}1 A%4](https://github.com/svenhb/GRBL-Plotter/assets/94118847/1d5d4b00-30ec-4b2f-9927-6a57138928fe)

**Comment by yifandianzi** [comment]
Hello, I just tried to turn on the Z-axis and disable the spindle like that, but the effect is still the same. I will release the pen

**Comment by svenhb** [comment]
Can you show the generated gcode

**Comment by yifandianzi** [comment]
Hello, is that the case This is preparing to write after zeroing
https://www.youtube.com/shorts/GNgGmXgAvh8

**Comment by svenhb** [comment]
I mean a screenshot like this
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/056ffc38-b758-4e4a-b2be-64e4046396a2)

**Comment by yifandianzi** [comment]
![~E2X} QZDDHYL2M3CDAZIH9](https://github.com/svenhb/GRBL-Plotter/assets/94118847/1d29799e-f465-4f15-bff7-7c227cafef7f)
In this way

**Comment by svenhb** [comment]
And the last lines of the code?

**Comment by svenhb** [comment]
If you watch my video https://youtu.be/tVS8A3_E2IY?t=141
the last action is, removing the pen, then end of program "M30" which causes the servo to move...

**Comment by yifandianzi** [comment]
![{A)A2}@JAZF9TEO$ON~TV`N](https://github.com/svenhb/GRBL-Plotter/assets/94118847/03904ddb-932d-410c-8992-615e02a12b5c)
Hello, it's from the 27-47 paragraph

**Comment by yifandianzi** [comment]
Execute M30 only after complete writing
![3Q4)OV7L}E~ J7E4P1U9P@M](https://github.com/svenhb/GRBL-Plotter/assets/94118847/2db02e8f-bb4e-462a-9b3f-62ef20e777ed)

**Comment by svenhb** [comment]
Try new version https://github.com/svenhb/GRBL-Plotter/blob/master/GRBL-Plotter_Setup_1_7_0_2_k.exe
It removes the final M05 in line 1354

**Comment by yifandianzi** [comment]
![image](https://github.com/svenhb/GRBL-Plotter/assets/94118847/28d8f84a-d8ae-4772-abae-bbd921c3b615)
Hello, my version you sent me will still have the same issue after installation

**Comment by svenhb** [comment]
The M05 should be gone:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/df204d8e-8c8b-4af9-beb0-225a95c5190e)

There is still M30 which is needed. You can try to edit the code mnually to see what happens:
Set edit mode:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/4d486ae2-5c92-46b7-9ab3-d8ef34bd7873)

remove last line, then update 2D-view:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/2d3d03b9-bb90-4aa1-ae74-e27a4a55f657)

**Comment by yifandianzi** [comment]
![image](https://github.com/svenhb/GRBL-Plotter/assets/94118847/e64187e5-dc77-4f5b-94cf-b6e171f7962f)
Hello, I am trying to delete M3 and the same problem will occur again

**Comment by yifandianzi** [comment]
https://www.youtube.com/watch?v=9yI0MYlkZ20
This is the process of starting work after resetting to zero, and it is from the process of the image that the servo starts to release the pen holder
![image](https://github.com/svenhb/GRBL-Plotter/assets/94118847/34b15d4e-1ffe-4fd2-9cfb-c99952125bcf)

**Comment by svenhb** [comment]
>Hello, I am trying to delete M3 and the same problem will occur again

The idea was to remove the final **M30** 

Your code still shows this M3 1000 - this controls your servo after pen change
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/8956b81c-a9c0-4937-a6c1-f7c9cbf6b1ec)

You must check this box!
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/57b946c6-12a3-41a2-bd85-d90aad614721)

**Comment by yifandianzi** [comment]
https://www.youtube.com/watch?v=zTcfYmqEp1U
Hello, here are my settings. Where else do I need to adjust them

**Comment by svenhb** [comment]
Sorry, I forget to check with option enabled "Add Tool change command".
I found a bug, please check https://github.com/svenhb/GRBL-Plotter/blob/master/GRBL-Plotter_Setup_1_7_0_2_l.exe

**Comment by yifandianzi** [comment]
Hello, when writing normally, what should I set for this Z-rise
https://www.youtube.com/watch?v=QgFRFkiDSZA

**Comment by svenhb** [comment]
When pen is in gripper, move down with joystick an press button "Zero  Z"

**Comment by svenhb** [comment]
BTW: you can set the custom buttons to perform the scripts (right click on desired button):
One button to select pen nr - press this first:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/a7277957-dff1-46df-8759-f4239c8ed696)

One button to start the script (set the path to your script:
![image](https://github.com/svenhb/GRBL-Plotter/assets/19973251/98956688-6d1c-40a9-81af-4dd77f07d30c)

**Comment by yifandianzi** [comment]
https://www.youtube.com/watch?v=6uB7rexIIOU
Thank you very much for your help. The plotter is now working, but the problem is that the pen clip is not tight enough, and the pen will come up when writing. I am trying to find a way to change the problem with the pen clip

**Comment by svenhb** [comment]
The problem is, you press the pen down and the pen can't evade. You may invert your mechanics and let the pen press down by the spring and lift by the servo.
My plotters uses gravity to move the pen down: https://youtu.be/7Nvf4UwK_yk?t=112
And evade by an extra linear guide https://youtu.be/XZIN8dHrgjg?t=11

**Comment by yifandianzi** [comment]
您好，请问我在归零的时候，Z只向上动一下就报错，是哪里设置不对吗，这是演示视频https://www.youtube.com/watch?v=tds28MmS-co

**Comment by svenhb** [comment]
Was your Z motor moving at all?
What is your $27 value? Alarm message mentions to increase pull-off setting ($27):
https://github.com/gnea/grbl/wiki/Grbl-v1.1-Configuration#27---homing-pull-off-mm

**Comment by yifandianzi** [comment]
![@JJSBW8IMB0CF %`91ZI$)Q](https://github.com/user-attachments/assets/d04dc663-fcbb-47c0-b102-66d6a0906e2b)

**Comment by svenhb** [comment]
Does this status changes, when you trigger the switch manually (by hand)? Pn should show a "Z" or not.
![image](https://github.com/user-attachments/assets/cf55b503-972e-4fe0-87c1-d57088fdff75)

**Comment by yifandianzi** [comment]
![$R17(RP75{YUAZL@AH53}_K](https://github.com/user-attachments/assets/2e92fdb1-7543-4c26-94c4-dd6d872842ca)
![LWBLHC%UA8_@USF6~{AW47B](https://github.com/user-attachments/assets/c4a52163-ad54-4e6e-8456-064f94bc2ac6)

**Comment by svenhb** [comment]
You need to seperate between "work" (big digits) and "machine" (small digits) coordinates:
![image](https://github.com/user-attachments/assets/ab234788-edc8-4c44-b155-12e73ad32086)

The origin of Work coordinates can be set with the buttons "Zero XY" etc.
The origin of machine coordinate will be found during "Home" and can not be changed.
Tool change position must be given in machine coordinates, to be independent of "zeroing" the origin.

I hope this helps: https://grbl-plotter.de/index.php?id=auto-pen-change&setlang=en

**Comment by yifandianzi** [comment]
Hello, this is the coordinate for resetting the machine to zero
![IMG_20241121_220342](https://github.com/user-attachments/assets/f539ec6c-ae6c-444a-9d4c-3722adbf32b5)
![PU{~YCY{6AMUA(6A_H1 BCB](https://github.com/user-attachments/assets/b55e2f66-8e84-428c-85ce-824dc5aa1cf3)

**Comment by svenhb** [comment]
So this is your configuration? And Position as shown in screen shot?
![image](https://github.com/user-attachments/assets/ece51462-39f0-4a69-a41f-bd634df5f361)

Then you can take over this values vor the shown machine area.
![image](https://github.com/user-attachments/assets/2410754f-a3a1-4947-855a-9fde7dacf18c)

**Comment by yifandianzi** [comment]
Yes, this is my configuration, X \ Y is like this, but after following your prompt and configuring it, it looks like this
![_}B~J6O VB%A{ SH(XN1E_I](https://github.com/user-attachments/assets/eedf26f7-0298-4c16-9933-5312ee5eb544)
![)G~P4$)2SZXJJY4_U)M)F~U](https://github.com/user-attachments/assets/bc653db6-913a-4145-bc2b-1587588a9566)

**Comment by svenhb** [comment]
You took the wrong numbers - you need the small numbers for machine coordinates
![image](https://github.com/user-attachments/assets/61bf9544-d72c-4d41-91d7-a36c95474623)

**Comment by svenhb** [comment]
Then you move to your tool change position, read the **machine** coordinate for use here:
![image](https://github.com/user-attachments/assets/9e2bb7b6-c83c-4a5a-8c73-096a7e1e927e)

**Comment by yifandianzi** [comment]
Hello, I've finished setting it up. Here's how it is
![~(_(~`E9 WT`)}5P$W`HORQ](https://github.com/user-attachments/assets/4985e778-69a5-4699-847b-57644075d255)
![L~F1 ~_IZPN1LZELMHP}2WY](https://github.com/user-attachments/assets/1b14ba34-1c6e-435e-906a-8b074d168fd9)

**Comment by svenhb** [comment]
Ok, now you need to adapt your pen holder positions:
![image](https://github.com/user-attachments/assets/48d56e26-b82c-4f0c-b46c-678cbb27ed23)

**Comment by yifandianzi** [comment]
Hello, I changed the X coordinate of the tool to 0, but it still shows in the common color area
![M%0PL){~{GM~HUNMO(8S0WO](https://github.com/user-attachments/assets/190810dc-c1dd-4246-b594-24565f2f8dfc)

**Comment by svenhb** [comment]
Just compare your screenshots - it moved to the right location.
Now change X to ca. 800 and Y to ca. 350:

![image](https://github.com/user-attachments/assets/5e3d7fda-6aac-4236-a610-6abbf07c472d)

**Comment by yifandianzi** [comment]
Hello, where is this not set correctly? When the fixture in front of the pen is moved after zeroing, Z is lifted to the limit position and then moved to the pen position to start gripping the pen. When writing, Z is almost lifted to the Z limit position. After writing, when placing the pen, Z is lifted again to trigger the Z limit position
https://www.youtube.com/watch?v=YUySwfs-eN8

**Comment by svenhb** [comment]
you need to set new origin for the Z work coordinate: move Z a little bit down (at least the "Save Z" value + the $27 value), then press "Zero Z".
![image](https://github.com/user-attachments/assets/9d55553d-f66c-42f1-8dd6-5e9aa9c6d6c5)

At the moment the work "0" is too close to the end switch.

Also, you may adapt $132 to the real mechanical dimensions, I assume Z can move 20 mm at all? But is not so important for the moment.

**Comment by svenhb** [comment]
Perhaps you changed something else?
Can you show your G-Code?
Which scripts do you use here - and did you modify them?
![image](https://github.com/user-attachments/assets/22a195b8-6535-410a-b582-976d66a11a9b)

**Comment by yifandianzi** [comment]
Hello, this is the script I am using
![8{ODWOELSNAT_3)6JDV1UFB](https://github.com/user-attachments/assets/3c9fcbea-8fb4-485d-827f-7286fb0b3897)
![A 8SU`Z3{32E)~@GPHME1N6](https://github.com/user-attachments/assets/150dcece-cc87-437b-88d4-2a0f56749625)
![QI9)MWPXA`XL88_@T$RL04L](https://github.com/user-attachments/assets/13e6e8c4-875c-45a4-860c-f9423d161dfb)

**Comment by yifandianzi** [comment]
![A 8SU`Z3{32E)~@GPHME1N6](https://github.com/user-attachments/assets/e700c50a-e88c-4495-b3c5-8ccbaec11a53)

**Comment by svenhb** [comment]
And the G-Code from the edit-box?

**Comment by yifandianzi** [comment]
Hello, the problem I am currently facing is: when selecting the pen for the first time, I can correctly grip it to write. However, when gripping the pen for the second time, there is an error in Z, and Z cannot fall to the specified undetermined value during writing. When gripping the pen for the third time, it cannot touch the paper at all. Can this offset value be adjusted

**Comment by svenhb** [comment]
So you have step loss on Z because your mechanical range is too small.
You may adapt Z movement in scripts to your needs:
![image](https://github.com/user-attachments/assets/4c1be0c9-39a6-4c30-ad76-3dfdc1393423)
![image](https://github.com/user-attachments/assets/45cba108-f8d7-4e08-b354-54dc48c8b823)

**Comment by yifandianzi** [comment]
Okay, thank you. Thank you for your help. I will continue to make adjustments

**Comment by yifandianzi** [comment]
Hello, I suddenly reported a malfunction while writing. Is it a problem with the driver
![1](https://github.com/user-attachments/assets/7ffff88e-6e49-4bf3-855e-a6541a144656)

**Comment by svenhb** [comment]
>ALARM:1

During your plotting process an end-switch was triggered. Perhaps your drawing is too big? 
You need to press "Kill Alarm".
You may disable $21 to avoid hard limit alarm. https://github.com/gnea/grbl/wiki/Grbl-v1.1-Configuration#21---hard-limits-boolean

**Comment by yifandianzi** [comment]
Thank you for your help

**Comment by yifandianzi** [comment]
Hello, I find this watercolor a bit interesting. Do you have any relevant information? I would like to try making it
![DW1WZR3K7Y{1HK6)T4Y AQG](https://github.com/user-attachments/assets/cc8f0412-f7d3-4a99-b61c-0b48eec4b2ec)

**Comment by svenhb** [comment]
It is my foldable plotter: http://svenhb.bplaced.net/?CNC___Klappbarer_Plotter
Together with some settings:

- Accurate PWM setup for different height positions
- Drag tool compensation for the brush
- Split original moves to insert water-color refresh script

![image](https://github.com/user-attachments/assets/087281b0-918b-4147-b0af-5d4a67db7ab8)

![image](https://github.com/user-attachments/assets/d7580e78-5561-4841-bb5e-38a5e4fc190d)

![image](https://github.com/user-attachments/assets/e069e67b-d477-498b-a9c7-f927c0069bf3)

**Comment by yifandianzi** [comment]
Okay, thank you. I'll refer to it

**Comment by yifandianzi** [comment]
Hello, is the sound emitted by the servo motor an electric current sound? How can I eliminate it
https://www.youtube.com/watch?v=prI3imqAP7Q

**Comment by svenhb** [comment]
This looks very good!!! I'm happy that you solved all problems!
You can speed up everything by:

- Use higher acceleration, perhaps 500 or 1000 for $120, 121
- Use higher XY-Feedrate in GRBL-Setup
- 

The servo noise is normal: the servo is trying to reach its target position when grabbing the pin, but the pin is in the way. Maybe adjust the grip position value a bit.

**Comment by svenhb** [comment]
For the gripper value try "S10" or "S20" or higher, you need to figure out:
![image](https://github.com/user-attachments/assets/eaf5e821-108c-41d7-a034-f1be654ef43f)

**Comment by yifandianzi** [comment]
Hello, I am working on a plotter but it is reporting an error. Is there something wrong
![b5ebde61f4d49eac9b985e1cb42e64a1](https://github.com/user-attachments/assets/4b9f8786-b39b-49ea-b469-040e485f1eb8)

**Comment by svenhb** [comment]
As the message said: during operation one of your end-switches was triggered.
You should increase the $27 value: https://github.com/gnea/grbl/wiki/Grbl-v1.1-Configuration#27---homing-pull-off-mm

**Comment by yifandianzi** [comment]
Hello, even after I changed the value to 27, it still looks like this
https://www.youtube.com/watch?v=5w1A4qMsxC4

**Comment by svenhb** [comment]
Your video shows: Z is already triggered and causes the alarm:
![image](https://github.com/user-attachments/assets/4a456cb2-0344-43ed-9821-4c170bd62d68)

**Comment by yifandianzi** [comment]
Yes, I have removed the Z limit line now, but it still shows the Z departure status

**Comment by yifandianzi** [comment]
Hello, after removing the Z limit, there was no fault reported. However, when I plugged in the mechanical limit switch, there was a fault reported
![IMG_20241229_171743](https://github.com/user-attachments/assets/9f389dd0-d990-4770-999b-0d62b5b3917f)

**Comment by svenhb** [comment]
I don't know the grbl firmware code, most probably there are several places where the limit switches will be checked.

**Comment by yifandianzi** [comment]
That's the firmware     https://github.com/cprezzi/grbl-servo

**Comment by Hostaler** [comment]
Hi, here's a picture of my limit switches for the Z.
My limit switches have three connections: Normally Open (NO), Normally Closed (NC), and Common or Signal (C). To avoid any errors with the limit switch (the same thing happened to me as to you), you have to connect C to NO. This way, I avoided that problem. If you find that it doesn't work that way, then connect C to NC.
It should work better for you, and on most motherboards, connect C to NO.
It doesn't matter if you connect +positive+ to one side or -negative- to the other, since it only tells the motherboard that there's a signal, and the signal flow doesn't matter.
I hope this helps.

![Image](https://github.com/user-attachments/assets/e0783c1e-1d3f-42b0-a733-9b8391a340a9)

**Comment by yifandianzi** [comment]
Hello, is there a problem with this limit switch? The test results are normal
https://www.youtube.com/shorts/Yx3lmrRgPoY
